server/controllers/booksController.ts

```diff
@@ -0,0 +1,126 @@
+import { Request, Response } from "express";
+import Book from "../models/bookModel";
+
+export const getBooks = async (req: Request, res: Response) => {
+  try {
+    const books = await Book.find();
+    res.status(200).json({
+      status: "success",
+      data: books,
+    });
+  } catch (err: any) {
+    res.status(500).json({
+      status: "error",
+      message: err.message,
+    });
+  }
+};
+
+export const saveBook = async (req: Request, res: Response): Promise<void> => {
+  try {
+    const bookData = req.body;
+    let existingBook = null;
+
+    if (bookData.id) {
+      existingBook = await Book.findOne({ id: bookData.id });
+    }
+
+    if (existingBook) {
+      res.status(200).json({
+        status: "success",
+        data: existingBook,
+        message: "Book already exists. Returning existing book.",
+      });
+      return;
+    }
+
+    const newBook = await Book.create(bookData);
+    res.status(201).json({
+      status: "success",
+      data: newBook,
+    });
+  } catch (err: any) {
+    res.status(400).json({
+      status: "error",
+      message: err.message,
+    });
+  }
+};
+
+export const deleteBook = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  try {
+    const deletedBook = await Book.findByIdAndDelete(req.params.id);
+    if (!deletedBook) {
+      res.status(404).json({
+        status: "error",
+        message: "Book not found",
+      });
+      return;
+    }
+    res.status(200).json({
+      status: "success",
+      data: null,
+    });
+  } catch (err: any) {
+    res.status(500).json({
+      status: "error",
+      message: err.message,
+    });
+  }
+};
+
+export const updateBook = async (
```

Comment:
I couldn't find this in the code when browsing - do you shield these endpoints against unauthorized user? Usually this is done by adding a middleware. Also, it's important to check for the ownership of the object as well (by e.g. keeping `ownerId` on the database object and disallowing changing it directly from the UI, only API could add it, apart from the ownership transfer cases).